core/src/main/java/com/linecorp/armeria/common/util/AsyncLoaderBuilder.java

```diff
@@ -0,0 +1,109 @@
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+/**
+ * A builder for creating a new {@link AsyncLoader}.
+ *
+ * <p>Expiration should be set by {@link #expireAfterLoad(Duration)} or {@link #expireIf(Predicate)}.
+ * If expiration is not set, {@link #build()} will throw {@link IllegalStateException}.
+ */
+@UnstableApi
+public final class AsyncLoaderBuilder<T> {
+
+    private final Function<@Nullable T, CompletableFuture<T>> loader;
+    @Nullable
+    private Duration expireAfterLoad;
+    @Nullable
+    private Predicate<? super T> expireIf;
+    @Nullable
+    private Predicate<? super T> refreshIf;
+    @Nullable
+    private BiFunction<? super Throwable, ? super @Nullable T,
+            ? extends @Nullable CompletableFuture<T>> exceptionHandler;
+
+    AsyncLoaderBuilder(Function<@Nullable T, CompletableFuture<T>> loader) {
+        requireNonNull(loader, "loader");
+        this.loader = loader;
+    }
+
+    /**
+     * Expires The loaded value after the duration since it was loaded.
+     * New value will be loaded by the loader function on next {@link AsyncLoader#get()}.
+     */
+    public AsyncLoaderBuilder<T> expireAfterLoad(Duration expireAfterLoad) {
+        requireNonNull(expireAfterLoad, "expireAfterLoad");
+        checkState(!expireAfterLoad.isNegative(), "expireAfterLoad: %s (expected: >= 0)", expireAfterLoad);
+        this.expireAfterLoad = expireAfterLoad;
+        return this;
+    }
+
+    /**
+     * Expires the loaded value if the predicate matches.
+     * New value will be loaded by the loader function on next {@link AsyncLoader#get()}.
+     */
+    public AsyncLoaderBuilder<T> expireIf(Predicate<? super T> expireIf) {
+        requireNonNull(expireIf, "expireIf");
+        this.expireIf = expireIf;
+        return this;
+    }
+
+    /**
+     * Asynchronously refreshes the loaded value which has not yet expired if the predicate matches.
+     * This pre-fetch strategy can remove an additional loading time on a cache miss.
+     */
+    public AsyncLoaderBuilder<T> refreshIf(Predicate<? super T> refreshIf) {
+        requireNonNull(refreshIf, "refreshIf");
+        this.refreshIf = refreshIf;
+        return this;
+    }
+
+    /**
+     * Handles the exception thrown by the loader function.
+     * If the exception handler returns {@code null}, {@link AsyncLoader#get()} completes exceptionally.
+     */
+    public AsyncLoaderBuilder<T> exceptionHandler(BiFunction<? super Throwable, ? super @Nullable T,
+            ? extends @Nullable CompletableFuture<T>> exceptionHandler) {
+        requireNonNull(exceptionHandler, "exceptionHandler");
+        this.exceptionHandler = exceptionHandler;
+        return this;
+    }
+
+    /**
+     * Returns a newly created {@link AsyncLoader} with the entries in this builder.
+     *
+     * @throws IllegalStateException if no expiration is set.
+     */
+    public AsyncLoader<T> build() {
+        if (expireAfterLoad == null && expireIf == null) {
```

Comment:
- We don't want to expose CacheEntry as public
- We want to provide 2 expire* method (`expireAfterLoad(duration), expireIf(predicate)` to user separately

Reply:
If you want to expose `CacheEntry`, we might consider exposing the cached time using `BiPredicate<T, Long>` rather than `CacheEntry`.

`CacheEntry` is designed for a specific implementation so I hesitated to expose it as a public API. If necessary, I would like to consider something more general like `TimestampedValue`.

Reply:
I see. I think it's fine to try to apply `AsyncLoader` to more use-cases and see if just the previous value is enough to determine expiration.

The API as-is can probably be improved later without too many breaking changes as well. Let's keep as-is for now.

Reply:
> The API as-is can probably be improved later without too many breaking changes as well

Agreed. The API needs some renovation after getting feedback from users.